File: shared/secrets/secrets.go
package secrets

import (
	"context"
	"fmt"
	"hash/crc32"
	"log"
	"os"

	secretmanager "cloud.google.com/go/secretmanager/apiv1"
	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
)

// GetSecret fetches a secret payload from Google Secret Manager.
// It accesses the "latest" version by default.
// It falls back to environment variables if the secret name exists as an env var.
func GetSecret(ctx context.Context, projectID, secretName string) (string, error) {
	// 1. Local Fallback
	if val := os.Getenv(secretName); val != "" {
		log.Printf("[SecretManager] Using local env var for: %s", secretName)
		return val, nil
	}

	client, err := secretmanager.NewClient(ctx)
	if err != nil {
		return "", fmt.Errorf("failed to create secretmanager client: %v", err)
	}
	defer client.Close()

	// Build the request.
	name := fmt.Sprintf("projects/%s/secrets/%s/versions/latest", projectID, secretName)
	req := &secretmanagerpb.AccessSecretVersionRequest{
		Name: name,
	}

	// Call the API.
	result, err := client.AccessSecretVersion(ctx, req)
	if err != nil {
		return "", fmt.Errorf("failed to access secret version: %v", err)
	}

	// Verify the data checksum.
	crc32c := crc32.MakeTable(crc32.Castagnoli)
	checksum := int64(crc32.Checksum(result.Payload.Data, crc32c))
	if result.Payload.DataCrc32C != nil && *result.Payload.DataCrc32C != checksum {
		return "", fmt.Errorf("datas corruption detected")
	}

	return string(result.Payload.Data), nil
}

File: shared/typescript/src/secrets/secrets.ts
import { SecretManagerServiceClient } from '@google-cloud/secret-manager';

let client: SecretManagerServiceClient;

function getClient() {
    if (!client) {
        client = new SecretManagerServiceClient();
    }
    return client;
}

/**
 * Fetches the latest version of a secret from Google Secret Manager.
 * @param projectId The Google Cloud Project ID.
 * @param secretName The name of the secret (not the full path).
 * @returns The secret string value.
 */
export async function getSecret(projectId: string, secretName: string): Promise<string> {
    // 1. Local Fallback: Check environment variable first
    // This allows local development without needing authenticated access to Secret Manager
    if (process.env[secretName]) {
        console.log(`[SecretManager] Using local env var for: ${secretName}`);
        return process.env[secretName]!;
    }

    // 2. Cloud Secret Manager
    const name = `projects/${projectId}/secrets/${secretName}/versions/latest`;

    try {
        const [version] = await getClient().accessSecretVersion({
            name: name,
        });

        const payload = version.payload?.data?.toString();
        if (!payload) {
            throw new Error(`Secret payload is empty for ${secretName}`);
        }
        return payload;
    } catch (err: any) {
        console.error(`Failed to fetch secret ${secretName}:`, err);
        throw err;
    }
}
